### ISSUE TITLE
torch.meshgrid is divergent from np.meshgrid

### ISSUE THREAD

**Issue by mruberry** [pytorch/pytorch#50276]
PyTorch's [torch.meshgrid](https://pytorch.org/docs/master/generated/torch.meshgrid.html?highlight=meshgrid#torch.meshgrid) is divergent from NumPy's [np.meshgrid](https://numpy.org/doc/stable/reference/generated/numpy.meshgrid.html?highlight=meshgrid#numpy.meshgrid). The documentation for these functions is very hard to parse, so I am not sure if this is a bug or not.

Snippet demonstrating the divergence:

```
nx, ny = (3, 2)
x = np.linspace(0, 1, nx)
y = np.linspace(0, 1, ny)
xv, yv = np.meshgrid(x, y)
xv
: array(
  [[0. , 0.5, 1. ], 
   [0. , 0.5, 1. ]])

xv, yv = torch.meshgrid(torch.from_numpy(x), torch.from_numpy(y))
xv
: tensor(
  [[0.0000, 0.0000],
   [0.5000, 0.5000],
   [1.0000, 1.0000]], dtype=torch.float64)
```

NumPy's np.meshgrid is [a top 100 NumPy function](https://github.com/Quansight-Labs/python-api-inspect/blob/master/data/csv/numpy-summary.csv). PyTorch's meshgrid is used hundreds of times within Facebook.

**Comment by kkduncan** [comment]
I also ran into a similar situation as a result of torch's meshgrid. It caused me to get very incorrect bounding boxes for an EfficientDet port that I was undertaking. `numpy.meshgrid(x, y) != torch.meshgrid(x, y)` but rather `torch.meshgrid(y, x)`. This should be fixed because it was a extremely hard bug to uncover. I wasted many hours behind it.

**Comment by mruberry** [comment]
Thank you for reporting this issue, @kkduncan, and I'm sorry it was so painful. User reports like yours are important to helping us prioritize fixes like this.

**Comment by rgommers** [comment]
This is the same as gh-15301. What's pointed out there is correct: numpy has an `indexing='xy'` keyword, `'ij'` mode was added later and this is what pytorch uses.

It's a little unfortunate that even if we add the `'xy'` mode to `torch.meshgrid` now, the defaults will be different. I don't see a good deprecation path to making the defaults the same. So what I think should be done:
1. Add a warning in the docs about `torch.meshgrid == numpy.meshgrid(..., indexing='ij')`
2. Implement `'xy'` mode for better numpy compatibility.

**Comment by rgommers** [comment]
Note: `indexing` is an often-used keyword. I would not recommend adding `copy=` or `sparse=`, those are less interesting.

**Comment by rgommers** [comment]
The plan in gh-55953 is to migrate to `indexing='xy'` over a period of 3 releases, to become compatible with NumPy and the array API standard.

**Comment by dagitses** [comment]
Thanks Ralf. That is the plan. The slightly more detailed plan is as-follows:

1) add an `OpInfo` for the operator. This is somewhat orthogonal, but it is related here in that it gives us the reference test as we approach conformance.
2) update meshgrid documentation to document the present discrepancies between the NumPy reference. Also, to clarify the operator's docstring to avoid confusion about what this function actually does.
3) add the indexing kwarg, only supporting the current `"ij"` behavior. Warn if the arg is not provided. For forward compatibility reasons, this requires adding a native function overload for two weeks.
4) implement support for `"xy"` indexing.
5) remove the temporary overload when ready.
6) after a PyTorch release contains the warning to use `indexing`, upgrade that warning to a runtime error
7) after a PyTorch release contains the runtime error requiring the use of `indexing`, change the default to `"xy"`

While 1/2 can be done in any order, this was arbitrarily linearized so that we may create a PR chain. Implicit aspects in each subsequent step are updates to the `OpInfo` test and docstring as appropriate.

**Comment by NicolasHug** [comment]
The current warning message is

> torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument.


As a small improvement, it would be helpful to add a small tip like

> Set indexing="ij" to preserve the default behavior and silence this warning.

This would avoid users to check the docs

**Comment by mruberry** [comment]
> The current warning message is
> 
> > torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument.
> 
> As a small improvement, it would be helpful to add a small tip like
> 
> > Set indexing="ij" to preserve the default behavior and silence this warning.
> 
> This would avoid users to check the docs

That's a great point, @NicolasHug (and we can say that indexing='xy' will be the future behavior, too). cc @dagitses. Should be a trivial fix.

**Comment by carlosgmartin** [comment]
`indexing='ij'` yields the most natural behavior. It also makes more sense in higher dimensions. `(N1, N2, N3,...Nn)` is more natural than `(N2, N1, N3,...Nn)`. I'm a bit dismayed that PyTorch is changing from the natural convention to numpy's unnatural convention.

**Comment by mruberry** [comment]
> `indexing='ij'` yields the most natural behavior. It also makes more sense in higher dimensions: `(N1, N2, N3, ..., Nn)` is more natural than `(N2, N1, N3,..., Nn)`. It's a little disappointing that PyTorch is changing from the natural behavior to NumPy's unnatural behavior, mimicking an arguably incorrect API choice.

It's just about user expectations and familiarity -- being different is a significant cost

**Comment by carlosgmartin** [comment]
@mruberry I understand. How about adding matlab's [ngrid](https://www.mathworks.com/help/matlab/ref/ndgrid.html) function, which has the former behavior? That seems like a good compromise.

**Comment by mruberry** [comment]
> @mruberry I understand. How about adding matlab's [ndgrid](https://www.mathworks.com/help/matlab/ref/ndgrid.html) function, which has the former behavior? That seems like a good compromise.

I'm sympathetic to this but also don't love adding a single matlab alias. How do you feel about defining it locally? It would be very simple.

> 
> It would be nice since `torch.ndgrid(x, y, z)` is more concise than `torch.meshgrid(x, y, z, indexing='ij')`, which I have to use a lot in my code.

Yeah I agree, although being explicit about the indexing mode may be more readable if anyone else is looking at your code.

**Comment by 99991** [comment]
# FOR THOSE LOOKING FOR HOW TO MIGRATE TO THE NEW API:
# `torch.meshgrid(x, y, indexing='ij')`

(Since I scrolled over [NicolasHug's](https://github.com/pytorch/pytorch/issues/50276#issuecomment-945860735) helpful response the first time and since this issue is the first google search result, I wrote it in a larger font since the error message still does not contain migration information. In this case, I could just have read `help(torch.meshgrid)` instead, but everyone goes for google search first anyway.)

```bash
.local/lib/python3.8/site-packages/torch/functional.py:445:
UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at  ../aten/src/ATen/native/TensorShape.cpp:2157.)
  return _VF.meshgrid(tensors, **kwargs)  # type: ignore[attr-defined]
```

**Comment by mruberry** [comment]
We should probably specify which indexing argument to use for meshgrid for those who want to preserve its behavior.

**Comment by mruberry** [comment]
Thanks for reporting this issue, @davidrs -- I'll take a look

**Comment by malfet** [comment]
@mruberry is there a  remaining work planned for 1.11? If not, can it be delimestoned?

**Comment by mruberry** [comment]
> @mruberry is there a remaining work planned for 1.11? If not, can it be de-milestoned?

No planned work for 1.11, no. I'll remove it from the milestone.

**Comment by vadimkantorov** [comment]
Also, torch meshgrid returns non-contig tensors while numpy returns contig:
```python
import numpy as np

xv, yv = np.meshgrid(np.arange(3), np.arange(2))
print(xv.flags['C_CONTIGUOUS'], yv.flags['C_CONTIGUOUS'])
# True True

import torch
xv, yv = torch.meshgrid(torch.arange(3), torch.arange(2), indexing = 'xy')
print(xv.is_contiguous(), yv.is_contiguous())
# False False
xv, yv = torch.meshgrid(torch.arange(3), torch.arange(2), indexing = 'ij')
print(xv.is_contiguous(), yv.is_contiguous())
# False False
```

A way of returning just a tensor (and not a tuple) should be good. Because currently it's not even contig... There might be some ways around (e.g. supporting idioms like `torch.stack(torch.meshgrid(...), dim = 0)` or `torch.stack(torch.meshgrid(...), dim = -1)`, but making them non-allocating and ensuring that the result is contig)

**Comment by justme220011** [comment]
C:\Users\OMEN\miniconda3\envs\avatarify\lib\site-packages\torch\functional.py:504: UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at C:\actions-runner\_work\pytorch\pytorch\builder\windows\pytorch\aten\src\ATen\native\TensorShape.cpp:3191.)
  return _VF.meshgrid(tensors, **kwargs)  # type: ignore[attr-defined]
Traceback (most recent call last):
  File "afy/cam_fomm.py", line 218, in <module>
    **predictor_args
  File "C:\Users\OMEN\avatarify-python-youtube\afy\predictor_local.py", line 49, in __init__
    self.fa = face_alignment.FaceAlignment(face_alignment.LandmarksType._2D, flip_input=True, device=self.device)
  File "C:\Users\OMEN\miniconda3\envs\avatarify\lib\enum.py", line 354, in __getattr__
    raise AttributeError(name) from None
AttributeError: _2D

somebopdy help me please , i am running avatarify-python-youtube, i deleted everything and download it again and this is what i am still getting please i really need help,
